File: algo-lib/1_basic/Python/LinkedQueue.py
# 使用LinkedList实现FIFO队列.

# 单个节点的表示.
class Node:
    empty = () # class attr. 用空的tuple表示empty
    # can be treated as sentinel.

    def __init__(self, item, next = empty):
        """ init a node """
        # 必须先检查next也是Node类的或者next是我们的empty.
        assert next is Node.empty or isinstance(next, Node)
        # 设定两个instance attr.
        self.item = item;
        self.next = next;

    def __repr__(self):
        """ 它实现Node类的repr()函数 """
        # 先处理next部分的~
        if self.next:
            # 如果next部分存在, 递归处理它的后面部分
            next_str = ', ' + repr(self.next)
        else:
            # 如果next部分不存在, 直接为空字符串就行
            next_str = ''
        return 'Node({0}{1})'.format(self.item, next_str)

class LinkedList:
    def __init__(self):
        node = Node(Node.empty)    # 创建第一个node 它作为哨兵.
        self.first = node;	  # 链表为空的状态是: 链表内只有一个哨兵node.
        self.last = node;
        self.N = 0;

    def isEmpty(self):
        return self.N == 0

    def enqueue(self, item):
        # 注意区分链表为空和非空的情况.
        if self.isEmpty():
            self.first.item = item
            self.N += 1
        else:
            node = Node(item)
            self.last.next = node
            self.last = node
            self.N += 1

    def dequeue(self):
        # same as pop
        if self.isEmpty():
            raise Exception('LinkedList instance is empty!')
        elif self.N == 1:
            value = self.first.item
            self.first.item = Node.empty
            self.N -= 1
            return value
        else:
            value = self.first.item
            self.first = self.first.next
            self.N -= 1
            return value

    def delete(self, item):     # @?@ not tested
        """ 删除从链表头遇到的第一个名为item的元素 """
        currentNode = self.first

        while currentNode.item != Node.empty:
            if currentNode.item == item:
                currentNode = currentNode.next
                return True
        return False

    def deleteRec(self, item):
        """ 递归版本的元素删除."""
        deleteRecCore(self, item, self.first)

    def deleteRecCore(self, item, node):
        if node == Node.empty:
            return False
        elif item == node.item:
            item = item.next
            return True
        else: 
            deleteRecCore(self, item, node.next)

    def iterate(self, item):
        """ 遍历并打印所有元素. """

if __name__ == "__main__":
    # test code.
    # 从标准输入流中读入一系列字符, 如果是-则pop, 如果是非- 则push进去.
    # usage: python LinkedList.py < tobe.txt

    import sys
    lst = LinkedList()
    # 从标准输入流读入字符.
    chars = sys.stdin.read().split() # get a chars list

    for c in chars:
        if c != "-":
            lst.enqueue(c)
        else:
            print(lst.dequeue())
    print("({0} left on LinkedList)".format(lst.N))

File: algo-lib/1_basic/Java/LinkedQueue.java
// String类型的队列. 内部数据结构为链表. 链表每一个元素item是String.

// 使用:
// java LinkedQueueOfStrings.java

import edu.princeton.cs.algs4.*;

public class LinkedQueue<Item> {
	private Node first, last;		// 链表头尾的指针
	private int N;					// 队列长度.

	// 构造函数省略了. first, last均会被初始化为默认值null.

	// Node representation
	private class Node {
		Item item;
		Node next;
	}

	public boolean isEmpty() {
		return first == null;
	}

	public int size() { return N; }

	public void enqueue(Item item) {
		Node oldlast = last;
		last = new Node();
		last.item = item;
		last.next = null;

		// 1. 原来链表就是空的.
		// 2. 其它情况
		if (isEmpty()) first = last;
		else oldlast.next = last;
		N++;
	}

	public Item dequeue() {
		Item item = first.item;
		first = first.next;
		if (isEmpty()) last = null;
		N--;
		return item;
		// note: 需要小心对空的queue进行出列操作. -> 这会直接抛出异常 所以还好吧.
		//       小心对只有一个node的queue进行出列操作.
	}

	// 需要考虑两个特殊状态: 
	// 1是刚刚建立好链表的状态 这个时候first, last都是null
	// 2是只有一个node的状态 此时first last都指向这个唯一的node.

	// 测试用例: 读取整数
	public static void main(String[] args) {
		LinkedQueue<String> q = new LinkedQueue<String>();

		while(!StdIn.isEmpty()) {
			String item = StdIn.readString();
			// 根据读取到的内容判断是出列还是入列.
			if (!item.equals("-"))
				q.enqueue(item);
			else if (!q.isEmpty()) StdOut.print(q.dequeue() + " ");
		}
		StdOut.println("(" + q.size() + " left on queue)");
	}
}